consensus/state.go

```diff
@@ -1286,6 +1287,13 @@ func (cs *State) defaultDoPrevote(height int64, round int32) {
 		return
 	}
 
+	if mustVoteNil {
```

Comment:
can you explain when this happens and why?

Reply:
Just realized, an alternate approach would be to add a check for `if cs.ValidBlock == nil`, and then make this gate the block being deemed valid.

Reply:
☝️ This would also make sense for me

Reply:
ah no wait `ValidBlock` depends on 2/3+ prevotes which wouldn't work.

I would instead add the check:
```
stateMachineValidBlock, err := cs.blockExec.ProcessProposal(cs.ProposalBlock)
if err != nil {
	cs.Logger.Error("state machine returned an error when trying to process proposal block", "err", err)
}
```
here as opposed to in `addProposalBlockPart()`. I also think we should be validating the block before getting the application to process it